### PR TITLE
Allow restricting profiles & profile_options based on JupyterHub groups

### PR DESCRIPTION
**Pull request by yuvipanda**
This PR does a bunch of things that would be difficult to do separately, and hence it's slightly big. Not too big!

# 1. Move from `Auth0OAuthenticator` to `GenericOAuthenticator` for EarthScope

The Auth0 authenticator made the EarthScope authentication mechanism look more custom than it really was - it's really just standard auth0, and if you look at the [auth0 authenticator](https://github.com/jupyterhub/oauthenticator/blob/main/oauthenticator/auth0.py) code it's fairly minimal - just a couple of convenience functions. This PR switches that (+ our auth0 documentation) to simply use GenericOAuthenticator. This has the following advantages:

1. The Auth0 documentation we have can be easily ported to just support any Generic OAuth provider if needed in the future, for other communities.
2. It signals that the EarthScope hub is not *that* special, just the first as a way for us to develop and offer new features. We should work on structuring how we do this, and signal when features are available in what hubs. But in the meantime, this reduces the overall apparent complexity to match actual complexity
3. Removes the custom logout_url work, and instead just mentions you need to set the `client_id` in the logout_url, and adds that to our auth0 documentation. This removes more custom code we have for EarthScope. Once https://github.com/jupyterhub/oauthenticator/pull/719 is merged, this helps us remove all custom code here from earthscope. This part fixes https://github.com/2i2c-org/infrastructure/issues/3715

There is still a custom override for the authenticator, but most of that can be removed once https://github.com/jupyterhub/oauthenticator/pull/719 is in a released version of OAuthenticator.

# 2. Provision JupyterHub groups for `scopes`

Auth0 allows us to use granted scopes to differentiate what rights various users on the JupyterHub should have. In this PR, we use our custom auth override to provision JupyterHub groups based on the scopes granted. We can not use the current `claim_groups_key`, as scopes are not a part of the user claim. https://github.com/jupyterhub/oauthenticator/pull/735 fixes this, but until that lands, we can rely on the custom auth override to provision JupyterHub groups.

# 3. `allowed_groups` functionality

We currently have `allowed_teams` functionality that restricts which profiles are available to which users based on GitHub team membership. However, this is specific to GitHubOAuthenticator only. This PR expands it into `allowed_groups` functionality, which works for *all* authenticators.  Until https://github.com/jupyterhub/oauthenticator/pull/735 lands, we also directly treat GitHub teams as 'groups'. All existing references to `allowed_teams` is renamed to `allowed_groups` with no loss of functionality for us. This also allows us to restrict profiles on EarthScope based on JupyterHub group membership.

# 4. `allowed_groups` support for `profile_option` 

Currently, `allowed_groups` (and `allowed_teams`) only supported restricting whole profiles (selected via radio buttons), but not profile_options (the dropdowns) or `unlisted_choice` (the textbox). This lead to some unclean duplication in places, and causes long term issues in places where the 'profile' is to select the environment and the dropdown is to select resource size. This PR adds `allowed_groups` support for both `profile_option.choices` and `profile_option.unlisted_choice`, so this awkwardness goes away.

Upstreaming strategy for this work is described in https://github.com/2i2c-org/infrastructure/issues/4021

### CODE REVIEW

**Comment by yuvipanda** [comment]
I'm going to mark this as draft, as I don't believe we should use our engineering capacity to review and merge this right now.

**Comment by yuvipanda** [comment]
This is missing additional feature docs on enabling `allowed_groups` in `profile_options`, but otherwise good to go.

**Comment by yuvipanda** [comment]
Added and cleaned up docs!

File: helm-charts/basehub/values.yaml [hunk 1]
```diff
@@ -993,109 +993,120 @@ jupyterhub:
...
+                        if 'allowed_groups' not in c:
+                          new_choices[k] = c
+                        # If `allowed_groups` *is* set for a profile option, it is allowed only for
+                        # members of that team.
+                        elif set(c['allowed_groups']) & groups:
+                          new_choices[k] = c
+                      po['choices'] = new_choices
+
+                # casefold teams so we can do case insensitive comparisons, as github itself is case insensitive (but preserving)
+                # for orgs and teams
```
Comment (GeorgianaElena) [inline helm-charts/basehub/values.yaml:1072]: I think this ended up here by accident and should instead be before line 1035 maybe?
```suggestion
```

**Review by GeorgianaElena — APPROVED** [review]
Thank you for working on this @yuvipanda , it's looking great! <3 I left a few small comments, but none is blocking so please feel free to merge this PR when you think it's ready 🚀

File: config/clusters/earthscope/common.values.yaml [hunk 2]
```diff
@@ -112,15 +89,15 @@ basehub:
...
           c.Spawner.auth_state_hook = populate_token
 
-          c.JupyterHub.authenticator_class = CustomAuth0OAuthenticator
+          c.JupyterHub.authenticator_class = CustomGenericOAuthenticator
       config:
-        JupyterHub:
-          authenticator_class: auth0
-        CustomAuth0OAuthenticator:
+        # JupyterHub:
+        #   authenticator_class: auth0
```
[the remark this answers is not in the document]
Comment (GeorgianaElena) [reply config/clusters/earthscope/common.values.yaml]: Old review comment remaining in the pipeline from a previous review attempt. It is fixed now <3

File: config/clusters/earthscope/common.values.yaml [hunk 3]
```diff
@@ -130,10 +115,14 @@ basehub:
...
             # https://2i2c.freshdesk.com/a/tickets/1280 for how this was granted.
             - geolab
+            - geolab:dev
+            - geolab:power
           extra_authorize_params:
             # This isn't an actual URL, just a string. Must not have a trailing slash
             audience: https://api.dev.earthscope.org
           username_claim: sub
+          # Convert 'scope' from the OAuth2 response into JupyterHub groups
+          manage_groups: true
```
[the remark this answers is not in the document]
Comment (yuvipanda) [reply config/clusters/earthscope/common.values.yaml:123]: I removed the line 100!

File: docs/howto/features/profile-list-restrict.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+All 2i2c engineers are part of the GitHub team `2i2c-org:hub-access-for-2i2c-staff`, so
+every `allowed_group` entry should have an explicit mention of that team so 2i2c engineers
+can access that option / profile and test it out when needed.
+
+## Enabling this feature for other Authenticators
+
+Currently, the EarthScope hub has this feature enabled via custom overrides. Once
+[this PR](https://github.com/jupyterhub/oauthenticator/pull/735) is merged and
+deployed, we can enable this feature for hubs using other Authenticators more generally.
```
[the remark this answers is not in the document]
Comment (yuvipanda) [reply docs/howto/features/profile-list-restrict.md:212]: Done

File: helm-charts/basehub/values.yaml (hunk 1 above)
Comment (yuvipanda) [reply helm-charts/basehub/values.yaml]: Good catch! Moved!

**Pull request merged**